homeassistant/components/flux_led/light.py

```diff
@@ -171,133 +187,226 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
-        scanner = BulbScanner()
-        await hass.async_add_executor_job(scanner.scan)
-
-        for device in scanner.getBulbInfo():
-            device_id = device["ipaddr"].replace(".", "_")
-            if device_id not in config_devices:
-                config_devices[device_id] = device
-
-    for device_id, device in config_devices.items():
-        add_device = {}
-        add_device["name"] = device.get("name", device_id)
-        add_device[CONF_HOST] = device[CONF_HOST]
-        add_device[CONF_PROTOCOL] = None
-        add_device[ATTR_MODE] = None
-        add_device[CONF_CUSTOM_EFFECT] = None
-        add_device[CONF_EFFECT_SPEED] = config_options.get(device_id, {}).get(
-            CONF_EFFECT_SPEED,
-            config_options.get("global", {}).get(
-                CONF_EFFECT_SPEED, DEFAULT_EFFECT_SPEED
-            ),
+
+    async def async_new_lights(bulbs: dict):
+        """Add new bulbs when they are found or configured."""
+
+        lights = []
+
+        for bulb_id, bulb_details in bulbs.items():
+            effect_speed = entry.options.get(bulb_id, {}).get(
+                CONF_EFFECT_SPEED,
+                entry.options.get("global", {}).get(
+                    CONF_EFFECT_SPEED, DEFAULT_EFFECT_SPEED
+                ),
+            )
+
+            coordinator = FluxLEDCoordinator(
+                hass=hass,
+                name=bulb_details[CONF_NAME],
+                update_interval=DEFAULT_SCAN_INTERVAL,
+                ip_address=bulb_details[CONF_HOST],
+            )
+
+            await coordinator.async_refresh()
+
+            lights.append(
+                FluxLight(
+                    coordinator=coordinator,
+                    unique_id=bulb_id,
+                    device=bulb_details,
+                    effect_speed=effect_speed,
+                )
+            )
+
+        async_add_entities(lights)
+
+    await async_new_lights(entry.data[CONF_DEVICES])
+
+    async_dispatcher_connect(hass, SIGNAL_ADD_DEVICE, async_new_lights)
+
+
+class FluxLEDCoordinator(DataUpdateCoordinator):
+    """Update Coordinator for a specific light entity."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        name: str,
+        update_interval: int,
+        ip_address: str,
+    ):
+        """Initialize the update coordinator."""
+
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=name,
+            update_interval=timedelta(seconds=update_interval),
         )
 
-        light = FluxLight(add_device)
-        lights.append(light)
+        self.light = WifiLedBulb(ip_address)
+
+    async def _async_update_data(self):
+        """Fetch the data from this light bulb."""
+
+        await self.hass.async_add_executor_job(self.light.update_state)
+
+        get_rgbw = await self.hass.async_add_executor_job(self.light.getRgbw)
+        get_rgb = await self.hass.async_add_executor_job(self.light.getRgb)
```

Comment:
Please make a function that calls all three calls above and then schedule that function just once on the executor. It's expensive to switch context.

Reply:
ok - think I got this one.

Reply:
Something like that, yes. The function can be a method too.

Reply:
Updated - think I got it.